File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  for (const mat of matrix) {
    const res = binarySearch(mat, target);
    if (res) {
      return true;
    }
  }
  return false;
};

const binarySearch = (matrix, target) => {
  let left = 0;
  let right = matrix.length - 1;

  if (target < matrix[left] || target > matrix[right]) {
    return false;
  }

  while (left <= right) {
    let mid = Math.floor((left + right) / 2);
    let midn = matrix[mid];

    if (midn === target) {
      return true;
    } else if (midn < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return false;
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution
{
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target)
    {
        for (auto &&mat : matrix)
        {
            bool res = binarySearch(mat, target);
            if (res)
            {
                return true;
            }
        }
        return false;
    }

    bool binarySearch(vector<int> &matrix, int target)
    {
        int left = 0;
        int right = matrix.size() - 1;
        if (target < matrix[left] || target > matrix[right])
        {
            return false;
        }

        while (left <= right)
        {
            int mid = (left + right) / 2;
            int midn = matrix[mid];
            if (target == midn)
            {
                return true;
            }
            else if (target < midn)
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }
        return false;
    }
};